File: rcc/services/steam_service.py
import os
from dataclasses import dataclass
from signal import Signals
from threading import Thread
import time

from psutil import NoSuchProcess, Process

from framework.logger import Logger
from rcc.communications.client.cmd.linux.mangohud_client import MANGO_HUD_CLIENT
from rcc.communications.client.cmd.linux.systemctl_client import SYSTEM_CTL_CLIENT
from rcc.communications.client.tcp.openrgb.effects.gaming import GAMING_EFFECT
from rcc.communications.client.websocket.steam.steam_client import STEAM_CLIENT
from rcc.models.gpu_brand import GpuBrand
from rcc.models.mangohud_level import MangoHudLevel
from rcc.models.settings import GameEntry
from rcc.models.wine_sync_option import WineSyncOption
from rcc.models.performance_profile import PerformanceProfile
from rcc.services.hardware_service import HARDWARE_SERVICE
from rcc.services.profile_service import PROFILE_SERVICE
from rcc.services.rgb_service import RGB_SERVICE
from rcc.utils.beans import EVENT_BUS
from rcc.utils.configuration import CONFIGURATION
from rcc.utils.constants import DEV_MODE, RCCDC_ASSET_PATH, USER_PLUGIN_FOLDER, USER_SCRIPTS_FOLDER
from rcc.utils.events import STEAM_SERVICE_CONNECTED, STEAM_SERVICE_DISCONNECTED, STEAM_SERVICE_GAME_EVENT
from rcc.utils.shell import SHELL


@dataclass
class RunningGameModel:
    """Running game class"""

    id: int
    name: str


class SteamService:
    """Steam service"""

    WRAPPER_PATH = os.path.join(USER_SCRIPTS_FOLDER, "wrapper")

    DECKY_SERVICE_PATH = os.path.expanduser(os.path.join("~", "homebrew", "services", "PluginLoader"))
    PLUGINS_FOLDER = os.path.expanduser(os.path.join("~", "homebrew", "plugins"))
    RCCDC_PATH = os.path.join(PLUGINS_FOLDER, "RCCDeckyCompanion")

    def __init__(self):
        self._logger = Logger()
        self._logger.info("Initializing SteamService")
        self._logger.add_tab()
        self._rccdc_enabled = False
        self.__running_games: dict[int, str] = {}
        self.__steam_connnected = STEAM_CLIENT.connected

        STEAM_CLIENT.on_connected(self.__on_steam_connected)
        STEAM_CLIENT.on_disconnected(self.__on_steam_disconnected)
        STEAM_CLIENT.on_launch_game(self.__launch_game)
        STEAM_CLIENT.on_stop_game(self.__stop_game)

        if STEAM_CLIENT.connected:
            self.__on_steam_connected(True)

        self._logger.rem_tab()

    def __on_steam_connected(self, on_boot=False):
        self.__running_games = {}
        self._logger.info("SteamClient connected")
        self._logger.add_tab()
        if not on_boot:
            self.__set_profile_for_games(True)
        self._logger.rem_tab()
        self.__steam_connnected = True
        EVENT_BUS.emit(STEAM_SERVICE_CONNECTED)

    def __on_steam_disconnected(self):
        self._logger.info("SteamClient disconnected")
        self.__steam_connnected = False
        if self.running_games:
            PROFILE_SERVICE.restore_profile()
        EVENT_BUS.emit(STEAM_SERVICE_DISCONNECTED)

    @property
    def steam_connected(self):
        """Flag for steam connection"""
        return self.__steam_connnected

    @property
    def metrics_enabled(self):
        """Flag metrics availability"""
        return MANGO_HUD_CLIENT.available

    def get_games(self) -> dict[int, GameEntry]:
        """Get games and setting"""
        return CONFIGURATION.games

    def __set_profile_for_games(self, on_connect=False):
        if len(self.__running_games) > 0:
            PROFILE_SERVICE.set_performance_profile(PerformanceProfile.PERFORMANCE, True, True)
        elif not on_connect:
            PROFILE_SERVICE.restore_profile()

        EVENT_BUS.emit(STEAM_SERVICE_GAME_EVENT, len(self.__running_games))

    def __launch_game(self, gid: int, name: str, pid: int):
        self._logger.info(f"Launched {name} with PID {pid}")
        self._logger.add_tab()

        if CONFIGURATION.games.get(gid) is None:
            self._logger.info("Game not configured, stopping...")
            process = Process(pid)
            env = process.environ()
            processes = process.children(recursive=True)
            processes.append(process)

            for child in reversed(processes):
                try:
                    child.send_signal(Signals.SIGKILL)
                except NoSuchProcess:
                    pass

            Thread(target=lambda: self.__first_game_run(gid, name, env)).start()
        elif gid not in self.running_games:
            self.__running_games[gid] = name
            HARDWARE_SERVICE.set_panel_overdrive(True)
            self.__set_profile_for_games()
            RGB_SERVICE.apply_effect(GAMING_EFFECT.name, True)

        self._logger.rem_tab()

    def __stop_game(self, gid: int, name: str):
        self._logger.info(f"Stopped {name}")
        pref_running_games = len(self.running_games)
        if gid in self.running_games:
            del self.running_games[gid]
        if pref_running_games != len(self.running_games):
            if len(self.running_games.keys()) == 0:
                RGB_SERVICE.restore_effect()
            self._logger.add_tab()
            HARDWARE_SERVICE.set_panel_overdrive(len(self.running_games) > 0)
            self.__set_profile_for_games()
            self._logger.rem_tab()

    def __first_game_run(self, gid: int, name: str, env: dict[str, str]):
        self._logger.info("Retrieving game details...")

        gpu = None
        if len(HARDWARE_SERVICE.gpus) > 0:
            for gpu_brand in GpuBrand:
                if gpu_brand in HARDWARE_SERVICE.gpus:
                    gpu = gpu_brand.value
                    break

        entry = GameEntry(name, gpu=gpu, proton="STEAM_COMPAT_PROTON" in env)
        details = STEAM_CLIENT.get_apps_details(gid)[0]
        launch_opts = details.launch_opts

        if "%command" in launch_opts:
            entry.env = launch_opts[0 : launch_opts.index("%command%")].strip()
            launch_opts = launch_opts[launch_opts.index("%command%") + 9 :].strip()

        if len(launch_opts) > 0:
            entry.args = launch_opts

        CONFIGURATION.games[gid] = entry
        CONFIGURATION.save_config()

        self._logger.info("Updating launch options...")
        STEAM_CLIENT.set_launch_options(gid, f"{self.WRAPPER_PATH} %command%")

        overlay_id = env["SteamOverlayGameId"]
        self._logger.info(f"Relaunching with Overlay GameId {overlay_id}...")
        time.sleep(1)
        SHELL.run_command(f"{env["STEAMSCRIPT"]} steam://rungameid/{overlay_id}")
        self._logger.info("Relauched")

    @property
    def running_games(self):
        """Get list of running games"""
        return self.__running_games

    @property
    def rccdc_enabled(self):
        """If plugin is enabled"""
        return self._rccdc_enabled

    def install_rccdc(self):
        """Install RCCDeckyCompanion plugin"""

        if os.path.exists(self.DECKY_SERVICE_PATH):
            if not os.path.exists(self.RCCDC_PATH):
                self._logger.info("Installing plugin for first time")
                self.__copy_plugin(RCCDC_ASSET_PATH, os.path.join(self.PLUGINS_FOLDER, "RCCDeckyCompanion"), False)
            else:
                if os.path.getmtime(self.RCCDC_PATH) < os.path.getmtime(RCCDC_ASSET_PATH):
                    self._logger.info("Updating Decky plugin")
                    self.__copy_plugin(RCCDC_ASSET_PATH, os.path.join(self.PLUGINS_FOLDER, "RCCDeckyCompanion"), True)
                else:
                    self._logger.debug("Plugin up to date")
            self._rccdc_enabled = True
        else:
            self._logger.warning("No Decky installation found, skipping plugin installation")
            self._rccdc_enabled = False

    def __copy_plugin(self, src: str, dst: str, is_update: bool):
        SHELL.run_command(f"cp -R {src} {USER_PLUGIN_FOLDER}", False)
        if is_update:
            SHELL.run_command(f"rm -R {dst}", True)
        SHELL.run_command(f"cp -R {os.path.join(USER_PLUGIN_FOLDER, 'RCCDeckyCompanion')} {dst}", True)
        Thread(target=lambda: SYSTEM_CTL_CLIENT.restart_service("plugin_loader")).start()

    def get_metrics_level(self, app_id) -> MangoHudLevel:
        """Get level for game"""
        game = CONFIGURATION.games.get(app_id)
        return (
            MangoHudLevel(game.metrics_level) if game and game.metrics_level is not None else MangoHudLevel.NO_DISPLAY
        )

    def set_metrics_level(self, metric_level: MangoHudLevel, app_id) -> MangoHudLevel:
        """Set level for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.metrics_level = metric_level.value if metric_level is not None else MangoHudLevel.NO_DISPLAY.value
        CONFIGURATION.save_config()

    def get_wine_sync(self, app_id) -> WineSyncOption:
        """Get level for game"""
        game = CONFIGURATION.games.get(app_id)
        return WineSyncOption(game.sync) if game and game.sync is not None else WineSyncOption.AUTO

    def set_wine_sync(self, wine_sync: WineSyncOption, app_id) -> MangoHudLevel:
        """Set level for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.sync = wine_sync.value if wine_sync is not None else WineSyncOption.AUTO.value
        CONFIGURATION.save_config()

    def get_prefered_gpu(self, app_id) -> GpuBrand | None:
        """Get GPU from game launch option"""
        game = CONFIGURATION.games.get(app_id)
        return GpuBrand(game.gpu) if game and game.gpu is not None else None

    def set_prefered_gpu(self, gpu_brand: GpuBrand, app_id) -> MangoHudLevel:
        """Set gpu for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.gpu = gpu_brand.value if gpu_brand is not None else None
        CONFIGURATION.save_config()

    def get_environment(self, app_id) -> str:
        """Get defined env for game"""
        game = CONFIGURATION.games.get(app_id)
        return game.env if game.env is not None else ""

    def set_environment(self, env: str, app_id) -> MangoHudLevel:
        """Set gpu for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.env = env
        CONFIGURATION.save_config()

    def get_parameters(self, app_id) -> str:
        """Get defined params for game"""
        game = CONFIGURATION.games.get(app_id)
        return game.args if game.args is not None else ""

    def is_proton(self, app_id) -> bool:
        """Check if app_id requires proton"""
        return CONFIGURATION.games.get(app_id).proton

    def set_parameters(self, params: str, app_id) -> MangoHudLevel:
        """Set gpu for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.args = params
        CONFIGURATION.save_config()

    def is_steamdeck(self, app_id):
        """Check if app_id simulates steamdeck"""
        return CONFIGURATION.games.get(app_id).steamdeck

    def set_steamdeck(self, steamdeck: bool, app_id) -> MangoHudLevel:
        """Set gpu for game launch option"""
        game = CONFIGURATION.games.get(app_id)
        game.steamdeck = steamdeck
        CONFIGURATION.save_config()

    def get_run_configuration(self, app_id):
        "Get environment and wrappers for app_id"
        environment = {}
        wrappers = []
        params = ""

        game = CONFIGURATION.games.get(app_id)
        if game is not None:
            environment["SteamDeck"] = "1" if game.steamdeck else "0"

            if game.proton:
                if DEV_MODE:
                    environment["PROTON_LOG"] = "1"

                mode = WineSyncOption(game.sync)
                if mode == WineSyncOption.NTSYNC:
                    environment["PROTON_NO_NTSYNC"] = "0"
                    environment["PROTON_NO_FSYNC"] = "1"
                    environment["PROTON_NO_ESYNC"] = "1"
                elif mode == WineSyncOption.FSYNC:
                    environment["PROTON_NO_NTSYNC"] = "1"
                    environment["PROTON_NO_FSYNC"] = "0"
                    environment["PROTON_NO_ESYNC"] = "1"
                elif mode == WineSyncOption.ESYNC:
                    environment["PROTON_NO_NTSYNC"] = "1"
                    environment["PROTON_NO_FSYNC"] = "1"
                    environment["PROTON_NO_ESYNC"] = "0"
                elif mode == WineSyncOption.NONE:
                    environment["PROTON_NO_NTSYNC"] = "1"
                    environment["PROTON_NO_FSYNC"] = "1"
                    environment["PROTON_NO_ESYNC"] = "1"

            if game.gpu is not None:
                environment.update(HARDWARE_SERVICE.get_gpu_selector_env(GpuBrand(game.gpu)))

            if self.metrics_enabled:
                environment["MANGOHUD_CONFIG"] = f"preset={game.metrics_level}"
                environment["MANGOHUD_DLSYM"] = "1"
                environment["MANGOHUD"] = "1"
                wrappers.append("mangohud")

            if game.env is not None and len(game.env.strip()):
                env = game.env.strip()
                parts = env.split(" ")
                for part in parts:
                    env_parts = part.split("=")
                    environment[env_parts[0]] = env_parts[1]

            if game.args is not None:
                params = game.args.strip()

        self._logger.info(f"Config for {game.name}:")
        self._logger.info(f"  Environment: {environment}")
        self._logger.info(f"  Wrappers: {wrappers}")

        return environment, wrappers, params


STEAM_SERVICE = SteamService()

File: RccCore/src/services/steam_service.cpp
#include "../../include/services/steam_service.hpp"

#include <iostream>
#include <set>
#include <sstream>

#include "../../include/configuration/configuration.hpp"
#include "../../include/models/hardware/gpu_brand.hpp"
#include "../../include/models/others/semantic_version.hpp"
#include "../../include/models/steam/steam_game_details.hpp"
#include "../../include/services/hardware_service.hpp"
#include "../../include/services/open_rgb_service.hpp"
#include "../../include/services/profile_service.hpp"
#include "RccCommons.hpp"

bool SteamService::metricsEnabled() {
	auto mangohud_which = shell.which("mangohud");
	return mangohud_which.has_value();
}

bool SteamService::isRunning(const unsigned int& appid) const {
	for (const auto& [key, val] : runningGames)
		if (key == appid)
			return true;
	return false;
}

const std::map<unsigned int, std::string>& SteamService::getRunningGames() const {
	return runningGames;
}

const std::map<std::string, GameEntry>& SteamService::getGames() {
	return configuration.getConfiguration().games;
}

SteamService::SteamService() {
	logger.info("Initializing SteamClient");
	logger.add_tab();

	SteamClient::getInstance();
	std::this_thread::sleep_for(std::chrono::milliseconds(25));
	if (SteamClient::getInstance().connected()) {
		onConnect(true);
	}

	installRccDC();

	SteamClient::getInstance().onConnect([this]() { onConnect(); });

	SteamClient::getInstance().onDisconnect([this]() { onDisconnect(); });

	SteamClient::getInstance().onGameLaunch([this](std::vector<std::any> data) {
		auto id	  = static_cast<uint32_t>(std::any_cast<int>(data[0]));
		auto name = std::any_cast<std::string>(data[1]);
		auto pid  = std::any_cast<int>(data[2]);

		onGameLaunch(id, name, pid);
	});

	SteamClient::getInstance().onGameStop([this](std::vector<std::any> data) {
		auto id	  = static_cast<uint32_t>(std::any_cast<int>(data[0]));
		auto name = std::any_cast<std::string>(data[1]);

		onGameStop(id, name);
	});

	logger.rem_tab();
}

void getPidsOfHierarchy(const pid_t parentId, std::set<pid_t>& pids) {
	for (auto& entry : std::filesystem::directory_iterator("/proc")) {
		std::string filename = entry.path().filename();
		if (entry.is_directory() && std::all_of(filename.begin(), filename.end(), ::isdigit)) {
			auto pid = std::stoi(filename);
			if (pids.find(pid) == pids.end()) {
				std::ifstream statFile("/proc/" + filename + "/stat");
				if (statFile.is_open()) {
					int pid_read, ppid;
					std::string comm, state;
					statFile >> pid_read >> comm >> state >> ppid;

					if (ppid == parentId) {
						pids.insert(pid);
						getPidsOfHierarchy(pid, pids);
					}
				}
			}
		}
	}
}

std::map<std::string, std::string> getProcessEnvironMap(pid_t pid) {
	std::map<std::string, std::string> env_map;
	std::string path = "/proc/" + std::to_string(pid) + "/environ";
	std::ifstream file(path, std::ios::in | std::ios::binary);

	if (!file.is_open()) {
		throw std::runtime_error("No se pudo abrir " + path);
	}

	std::string buffer((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

	size_t start = 0;
	while (start < buffer.size()) {
		size_t end = buffer.find('\0', start);
		if (end == std::string::npos)
			break;
		std::string entry = buffer.substr(start, end - start);
		size_t eq_pos	  = entry.find('=');
		if (eq_pos != std::string::npos) {
			std::string key	  = entry.substr(0, eq_pos);
			std::string value = entry.substr(eq_pos + 1);
			env_map[key]	  = value;
		}
		start = end + 1;
	}

	return env_map;
}

void SteamService::onFirstGameRun(unsigned int gid, std::string name, std::map<std::string, std::string> environment) {
	logger.info("Configuring game");
	logger.add_tab();

	std::optional<std::string> gpu = std::nullopt;
	if (hardwareService.getGpus().size() > 0) {
		auto items = GpuBrand::getAll();
		auto gpus  = hardwareService.getGpus();
		for (GpuBrand g : GpuBrand::getAll()) {
			if (gpus.find(g.toString()) != gpus.end()) {
				gpu = g.toString();
				break;
			}
		}
	}

	auto proton = environment.find("STEAM_COMPAT_PROTON") != environment.end();

	SteamGameDetails details = SteamClient::getInstance().getAppsDetails({gid})[0];
	auto launch_opts		 = details.launch_opts;

	std::string env, args;

	const std::string marker = "%command%";

	size_t pos = launch_opts.find(marker);
	if (pos != std::string::npos) {
		env = launch_opts.substr(0, pos);

		env.erase(0, env.find_first_not_of(" \t\n\r"));
		env.erase(env.find_last_not_of(" \t\n\r") + 1);

		launch_opts = launch_opts.substr(pos + marker.size());
		launch_opts.erase(0, launch_opts.find_first_not_of(" \t\n\r"));
		launch_opts.erase(launch_opts.find_last_not_of(" \t\n\r") + 1);
	}
	if (!launch_opts.empty()) {
		args = launch_opts;
	}

	auto overlayId = environment.find("SteamOverlayGameId")->second;

	GameEntry entry{args, env, gpu, MangoHudLevel::Enum::NO_DISPLAY, name, overlayId, proton, false, WineSyncOption::Enum::AUTO};
	configuration.getConfiguration().games[std::to_string(gid)] = entry;
	configuration.saveConfig();

	SteamClient::getInstance().setLaunchOptions(gid, WRAPPER_PATH + " %command%");
	logger.info("Configuration finished");

	logger.info("Relaunching game with SteamOverlayId " + overlayId + "...");
	std::this_thread::sleep_for(std::chrono::milliseconds(250));

	shell.run_command("steam steam://rungameid/" + overlayId);

	logger.rem_tab();
	logger.rem_tab();
}

bool SteamService::checkIfRequiredInstallation() {
	std::ifstream fileRunning(Constants::RCCDC_ASSET_PATH + "/package.json");

	nlohmann::json jR;
	fileRunning >> jR;

	SemanticVersion vA = SemanticVersion::parse(Constants::PLUGIN_VERSION);
	SemanticVersion vR = SemanticVersion::parse(jR["version"]);

	return vA > vR;
}

void SteamService::installRccDC() {
	try {
		if (FileUtils::exists(Constants::DECKY_SERVICE_PATH)) {
			if (!FileUtils::exists(Constants::RCCDC_PATH)) {
				logger.info("Installing plugin for first time");
				copyPlugin();
			} else {
				if (checkIfRequiredInstallation()) {
					logger.info("Updating Decky plugin");
					copyPlugin();
				} else {
					logger.info("Plugin up to date");
				}
			}
			rccdcEnabled = true;
		} else {
			logger.warn("No Decky installation found, skipping plugin installation");
			rccdcEnabled = false;
		}
	} catch (std::exception e) {
		logger.error("Error while installing RCCDeckyCompanion plugin: {}", e.what());
	}
}

void SteamService::copyPlugin() {
	installer = std::thread([this]() {
		FileUtils::copy(Constants::RCCDC_ASSET_PATH, Constants::USER_PLUGIN_DIR);

		std::vector<std::string> commands{"if [[ -d \"" + Constants::RCCDC_PATH + "\" ]]; then rm -R \"" + Constants::RCCDC_PATH + "\"; fi",
										  "chmod 777 \"" + Constants::PLUGINS_FOLDER + "/..\"",
										  "mv \"" + Constants::USER_PLUGIN_DIR + "\" \"" + Constants::RCCDC_PATH + "\"",
										  "systemctl restart plugin_loader.service"};

		for (auto cmd : commands) {
			shell.run_elevated_command(cmd);
		}
	});
}

void SteamService::onGameLaunch(unsigned int gid, std::string name, int pid) {
	logger.info("Launched '" + name + "' (" + std::to_string(gid) + ") with PID " + std::to_string(pid));
	logger.add_tab();
	if (configuration.getConfiguration().games.find(std::to_string(gid)) == configuration.getConfiguration().games.end()) {
		logger.info("Game not configured");
		logger.add_tab();

		logger.info("Getting process environment...");
		logger.add_tab();
		auto env = getProcessEnvironMap(pid);
		logger.rem_tab();

		logger.info("Stopping process...");
		logger.add_tab();
		std::set<pid_t> pids;
		pids.insert(pid);
		getPidsOfHierarchy(pid, pids);

		std::ostringstream oss;
		oss << "kill -9";
		for (pid_t pid : pids) {
			oss << " " << pid;
		}
		std::this_thread::sleep_for(std::chrono::milliseconds(250));
		shell.run_elevated_command(oss.str(), false);
		logger.rem_tab();

		std::thread([this, gid, name, env]() { onFirstGameRun(gid, name, env); }).detach();
		;
	} else if (runningGames.find(gid) == runningGames.end()) {
		runningGames[gid] = name;
		setProfileForGames();

		eventBus.emit_event(Events::STEAM_SERVICE_GAME_EVENT, {runningGames.size()});
	}
	logger.rem_tab();
}

void SteamService::onGameStop(unsigned int gid, std::string name) {
	if (runningGames.find(gid) != runningGames.end()) {
		logger.info("Stopped '" + name + "' (" + std::to_string(gid) + ")");
		runningGames.erase(gid);
		{
			logger.add_tab();
			setProfileForGames();
			logger.rem_tab();

			eventBus.emit_event(Events::STEAM_SERVICE_GAME_EVENT, {runningGames.size()});
		}
	}
}

void SteamService::setProfileForGames(bool onConnect) {
	if (!runningGames.empty()) {
		hardwareService.setPanelOverdrive(true);
		openRgbService.setEffect("Gaming", true);
		profileService.setPerformanceProfile(PerformanceProfile::Enum::PERFORMANCE, true, true);
	} else if (!onConnect) {
		hardwareService.setPanelOverdrive(false);
		openRgbService.restoreAura();
		profileService.restoreProfile();
	}
}

void SteamService::onConnect(bool onBoot) {
	logger.info("Connected to Steam");
	logger.add_tab();
	if (!onBoot) {
		setProfileForGames(true);
	}
	eventBus.emit_event(Events::STEAM_SERVICE_GAME_EVENT, {runningGames.size()});
	logger.rem_tab();
}

void SteamService::onDisconnect() {
	logger.info("Disconnected from Steam");
	logger.add_tab();
	if (!runningGames.empty()) {
		profileService.restoreProfile();
		openRgbService.restoreAura();
		runningGames.clear();
	}
	logger.rem_tab();
}

const SteamGameConfig SteamService::getConfiguration(const std::string& gid) {
	SteamGameConfig cfg;

	auto games = configuration.getConfiguration().games;
	auto it	   = games.find(gid);

	std::optional<GameEntry> entry = std::nullopt;
	if (it != games.end()) {
		entry = it->second;
	} else {
		for (const auto& [key, val] : configuration.getConfiguration().games) {
			if (val.overlayId == gid) {
				entry = val;
				break;
			}
		}
	}

	if (entry.has_value()) {
		auto gameEntry = entry.value();
		if (gameEntry.env.has_value()) {
			std::istringstream ss(gameEntry.env.value());
			std::string token;
			std::map<std::string, std::string> env_vars;

			while (ss >> token) {
				auto pos = token.find('=');
				if (pos != std::string::npos) {
					std::string key	  = token.substr(0, pos);
					std::string value = token.substr(pos + 1);
					env_vars[key]	  = value;
				}
			}

			for (const auto& [key, val] : env_vars) {
				cfg.environment[key] = val;
				logger.info("Added env {}={}", key, val);
			}
		}
		if (gameEntry.args.has_value()) {
			cfg.parameters = gameEntry.args.value();
		}

		cfg.environment["SteamDeck"] = gameEntry.steamdeck ? "1" : "0";
		if (gameEntry.proton) {
			cfg.environment["PROTON_USE_NTSYNC"] = gameEntry.sync == WineSyncOption::Enum::NTSYNC ? "1" : "0";
			cfg.environment["PROTON_NO_NTSYNC"]	 = gameEntry.sync == WineSyncOption::Enum::NTSYNC ? "0" : "1";
			cfg.environment["PROTON_NO_FSYNC"]	 = gameEntry.sync == WineSyncOption::Enum::FSYNC ? "0" : "1";
			cfg.environment["PROTON_NO_ESYNC"]	 = gameEntry.sync == WineSyncOption::Enum::ESYNC ? "0" : "1";
		}

		if (gameEntry.gpu.has_value()) {
			auto gpuEnv = hardwareService.getGpuSelectorEnv(gameEntry.gpu.value());
			for (const auto& [key, val] : gpuEnv) {
				cfg.environment[key] = val;
			}
		}

		auto mangohud_which = shell.which("mangohud");
		if (mangohud_which.has_value()) {
			cfg.environment["MANGOHUD_CONFIG"] = "preset=" + std::to_string(gameEntry.metrics_level.toInt());
			cfg.environment["MANGOHUD_DLSYM"]  = "1";
			cfg.environment["MANGOHUD"]		   = "1";
			cfg.wrappers.emplace_back(mangohud_which.value());
		}
	} else {
		logger.error("No configuration entry found for " + gid);
	}

	return cfg;
}

const std::optional<GpuBrand> SteamService::getPreferedGpu(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].gpu;
}

void SteamService::setPreferedGpu(const unsigned int& gid, const std::optional<GpuBrand>& gpu) {
	if (gpu.has_value())
		configuration.getConfiguration().games[std::to_string(gid)].gpu = gpu.value().toString();
	else
		configuration.getConfiguration().games[std::to_string(gid)].gpu = std::nullopt;

	configuration.saveConfig();
}

bool SteamService::isSteamDeck(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].steamdeck;
}

void SteamService::setSteamDeck(const unsigned int& gid, const bool& value) {
	configuration.getConfiguration().games[std::to_string(gid)].steamdeck = value;
	configuration.saveConfig();
}

const MangoHudLevel SteamService::getMetricsLevel(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].metrics_level;
}

void SteamService::setMetricsLevel(const unsigned int& gid, const MangoHudLevel& level) {
	configuration.getConfiguration().games[std::to_string(gid)].metrics_level = level;
	configuration.saveConfig();
}

const WineSyncOption SteamService::getWineSync(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].sync;
}

void SteamService::setWineSync(const unsigned int& gid, const WineSyncOption& level) {
	configuration.getConfiguration().games[std::to_string(gid)].sync = level;
	configuration.saveConfig();
}

bool SteamService::isProton(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].proton;
}

const std::string SteamService::getEnvironment(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].env.value_or("");
}

void SteamService::setEnvironment(const unsigned int& gid, const std::string& env) {
	configuration.getConfiguration().games[std::to_string(gid)].env = env;
	configuration.saveConfig();
}

const std::string SteamService::getParameters(const unsigned int& gid) {
	return configuration.getConfiguration().games[std::to_string(gid)].args.value_or("");
}

void SteamService::setParameters(const unsigned int& gid, const std::string& args) {
	configuration.getConfiguration().games[std::to_string(gid)].args = args;
	configuration.saveConfig();
}